feat(monitoring): automate statuspage and new relic integration (WIP)

rofe opened the pull request:
Fix #46

WIP commit to validate chosen approach:
- 2 separate tools for statuspage and new relic due to number and naming of arguments (separate auth keys etc). it currently works like this (see `README.md` for detailed description):
  - run `statuspage create` (this already works, also with `--group_id`)
  - copy automation email in console output
  - run `newrelic create <url> <email> (this currently only creates a new relic monitor, as already implemented in `bin.js`)

TODO:
- [x] implement creation of notification channel
- [x] implement creation/update of alert policy
- [x] delete `bin.js`
- [x] change circleci config to use new tools

trieloff commented on README.md:
```diff
@@ -145,38 +145,86 @@ $http.get('https://adobeioruntime.net/api/v1/web/helix/helix-services/status@v3/
-$ npx synthetics-check
-synthetics-check <cmd> url
+$ npx newrelic
+newrelic-synthetics <cmd> url
```
This looks inconsistent, it should return the name of the command entered above.

rofe replied:
Good catch. I changed tool names back and forth but eventually settled for simplified versions: `statuspage` and `newrelic`

trieloff commented on package.json:
```diff
@@ -4,6 +4,8 @@
   "bin": {
+    "newrelic": "src/newrelic.js",
+    "statuspage": "src/statuspage.js",
     "synthetics-check": "src/bin.js"
```
This still needs to go.

rofe replied:
👍  see TODO above :)

trieloff commented on src/newrelic.js:
```diff
@@ -0,0 +1,200 @@
+      return [];
+    }
+  } catch (e) {
+    console.error('Unable to retrieve monitors');
```
Please check the log changes from @koraa's other branch.

rofe replied:
I'll wait for her branch to be merged 😄 

Although in the case of this type of tooling, i would argue that you expect to see the output in the `console`, no?

rofe commented:
Update: after a call with New Relic support, "Alerts Location Failure Conditions" is now enabled in our account, so we should now be able to recreate these types of conditions via REST API.

rofe commented:
Quite a few changes went into the last commit:
- using the desired condition type now, with multiple location failures and threshold of 2
- simplified the command line tools to accept a single `setup` command
- reduced the number of default locations from 18 to 7 (request from New Relic/Adobe liaison)
- added support for an optional "group alert policy" (e.g. "Delivery Repeated Failure") where the monitor will be added in addition to its dedicated alert policy/condition

rofe commented:
@trieloff coud you take another look?

rofe commented on .circleci/config.yml:
```diff
@@ -81,7 +81,7 @@ jobs:
     - run:
         name: Set Up Monitoring
-        command: node src/bin.js update https://adobeioruntime.net/api/v1/web/helix/helix-services/`node -e "console.log(require('./package.json').name.replace('@adobe/helix-', ''))"`@v`node -e "console.log(require('./package.json').version.match(/^[0-9]+.[0-9]+/)[0])"`/_status_check/healthcheck.json
+        command: node src/newrelic.js setup https://adobeioruntime.net/api/v1/web/helix/helix-services/`node -e "console.log(require('./package.json').name.replace('@adobe/helix-', ''))"`@v`node -e "console.log(require('./package.json').version.match(/^[0-9]+.[0-9]+/)[0])"`/_status_check/healthcheck.json `node src/statuspage.js setup --silent`
```
@trieloff: I'm not sure if the `v` is correct here, helix services seem to get deployed without it, e.g.
```
https://adobeioruntime.net/api/v1/web/helix/helix-services/status@4.3.5/_status_check/healthcheck.json
```

trieloff replied:
`@vX` and `@vX.Y` is for the major/minor symlinks.

rofe commented:
The new implementation can be tested by running the following stacked command from the `automate-statuspage-newrelic` branch:
```bash
$ node src/newrelic.js setup https://adobeioruntime.net/api/v1/web/helix/helix-services/`node -e "console.log(require('./package.json').name.replace('@adobe/helix-', ''))"`@v`node -e "console.log(require('./package.json').version.match(/^[0-9]+.[0-9]+/)[0])"`/_status_check/healthcheck.json \`node src/statuspage.js setup --silent \
--name "My Test" \
--auth <your_statuspage_key>` \
--name "My Test" \
-- auth <newrelic_admin_key>
```
The `--name` option in both commands will make sure you won't override any existing component, monitor or alert policy.

trieloff commented:
> reduced the number of default locations from 18 to 7 (request from New Relic/Adobe liaison)

Is this related to #50 ?

trieloff commented on README.md:
```diff
@@ -145,38 +145,100 @@ $http.get('https://adobeioruntime.net/api/v1/web/helix/helix-services/status@v3/
+`newrelic` automates the following New Relic features:
+1. creation or update of monitors in New Relics Synthetics
+1. creation of notification channels in New Relic Alerts
+1. creation or update of alert policies in New Relic Alerts
```
@rofe you mentioned that New Relic had to enable an API for our account to make this work. It would be worth documenting this here or below.

rofe replied:
Sure. It's documented here: https://docs.newrelic.com/docs/multi-location-synthetics-alert-conditions

I'll add the link to the readme.

trieloff commented on src/newrelic/synthetics.js:
```diff
@@ -0,0 +1,136 @@
+const MONITOR_FREQUENCY = 15;
+const MONITOR_STATUS = 'ENABLED';
+const MONITOR_THRESHOLD = 7;
+const MONITOR_LOCATIONS = [
```
I don't know the full story behind this, but should we make this configurable (maybe in a separate PR). Should we reduce the interval to make up for the reduction in locations?

rofe replied:
Readded all AWS locations after confirming with New Relic.

trieloff commented on src/newrelic/synthetics.js:
```diff
@@ -0,0 +1,136 @@
+  'AWS_EU_WEST_2',
+  'AWS_SA_EAST_1',
+  'AWS_US_EAST_1',
+  'AWS_US_WEST_1',
```
Also (not something you can fix in the PR), but worth discussing: why are there only AWS locations? If there is an AWS<->World connectivity problem, our monitors aren't going to catch it because Runtime is on AWS, too.

rofe replied:
True, but my understanding is that New Relic deprecated all of the non AWS locations.

trieloff replied:
It's more a question for New Relic than for you.

rofe replied:
According to NR, all locations are currently AWS, but there's the option to add private minions (not available as a service yet).
